flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNHighAvailabilityITCase.java

```diff
@@ -18,198 +18,205 @@
-							try {
-								ActorGateway gateway = LeaderRetrievalUtils.retrieveLeaderGateway(
-									finalHighAvailabilityServices.getJobManagerLeaderRetriever(HighAvailabilityServices.DEFAULT_JOB_ID),
-									actorSystem,
-									timeout);
-								ActorGateway selfGateway = new AkkaActorGateway(getRef(), gateway.leaderSessionID());
-
-								gateway.tell(new TestingJobManagerMessages.NotifyWhenAtLeastNumTaskManagerAreRegistered(1), selfGateway);
-
-								expectMsgEquals(Acknowledge.get());
-
-								gateway.tell(PoisonPill.getInstance());
-							} catch (Exception e) {
-								throw new AssertionError("Could not complete test.", e);
-							}
-						}
-					};
-				}
-
-				new Within(timeout) {
-					@Override
-					protected void run() {
-						try {
-							ActorGateway gateway = LeaderRetrievalUtils.retrieveLeaderGateway(
-								finalHighAvailabilityServices.getJobManagerLeaderRetriever(HighAvailabilityServices.DEFAULT_JOB_ID),
-								actorSystem,
-								timeout);
-							ActorGateway selfGateway = new AkkaActorGateway(getRef(), gateway.leaderSessionID());
-
-							gateway.tell(new TestingJobManagerMessages.NotifyWhenAtLeastNumTaskManagerAreRegistered(1), selfGateway);
-
-							expectMsgEquals(Acknowledge.get());
-						} catch (Exception e) {
-							throw new AssertionError("Could not complete test.", e);
-						}
-					}
-				};
-
-			}};
-		} finally {
-			if (yarnClusterClient != null) {
-				log.info("Shutting down the Flink Yarn application.");
-				yarnClusterClient.shutDownCluster();
-				yarnClusterClient.shutdown();
-			}
-
-			if (highAvailabilityServices != null) {
-				highAvailabilityServices.closeAndCleanupAllData();
-			}
+	public void testKillYarnSessionClusterEntrypoint() throws Exception {
+		assumeTrue(
+			"This test kills processes via the pkill command. Thus, it only runs on Linux, Mac OS, Free BSD and Solaris.",
+			OperatingSystem.isLinux() || OperatingSystem.isMac() || OperatingSystem.isFreeBSD() || OperatingSystem.isSolaris());
+
+		final YarnClusterDescriptor yarnClusterDescriptor = setupYarnClusterDescriptor();
+
+		final RestClusterClient<ApplicationId> restClusterClient = deploySessionCluster(yarnClusterDescriptor);
+
+		final JobGraph job = createJobGraph();
+
+		final JobID jobId = submitJob(restClusterClient, job);
+
+		final ApplicationId id = restClusterClient.getClusterId();
+
+		final long retryTimeout = 100L;
+		waitUntilJobIsRunning(restClusterClient, jobId, retryTimeout);
+
+		killApplicationMaster(yarnClusterDescriptor.getYarnSessionClusterEntrypoint());
+
+		final YarnClient yarnClient = getYarnClient();
+		Assert.assertNotNull(yarnClient);
+
+		while (yarnClient.getApplicationReport(id).getCurrentApplicationAttemptId().getAttemptId() < 2) {
+			Thread.sleep(retryTimeout);
+		}
+
+		waitUntilJobIsRunning(restClusterClient, jobId, retryTimeout);
+
+		yarnClient.killApplication(id);
+
+		while (yarnClient.getApplications(EnumSet.of(YarnApplicationState.KILLED)).size() == 0 &&
```

Comment:
Wouldn't this be semantically equivalent?
```suggestion
		while (yarnClient.getApplications(EnumSet.of(YarnApplicationState.KILLED, YarnApplicationState.FINISHED)).isEmpty()) {
```

Reply:
Yes this is a good improvement. Will adapt it.